Added Dark Theme, Monospace font and Font size options

CieNTi opened the pull request:
I have tested only on Linux Debian stretch under Qt Creator (see about capture below):

![about qt creator_225](https://user-images.githubusercontent.com/5205702/36127639-c45b98bc-105e-11e8-960a-8d13a0ab0480.png)

Defaults

![cutecom - default_226](https://user-images.githubusercontent.com/5205702/36127652-ce635084-105e-11e8-92e6-cbb2d89c2d42.png)

Dark + Monospace + Font size changed

![cutecom - default_227](https://user-images.githubusercontent.com/5205702/36127655-d1836830-105e-11e8-81b1-141b8d6d53aa.png)

Hex view

![cutecom - default_228](https://user-images.githubusercontent.com/5205702/36127656-d455b0d6-105e-11e8-9795-d41244d75a2b.png)

Changes are not reflected until CuteCom is restarted.

Please tell me if something is wrong or a tweak is needed,

Regards

CieNTi committed:
Added ColinDuquesnoy dark theme v2.5.1 option to settings and cpanel

CieNTi committed:
Load dark theme at boot time

CieNTi committed:
Added setDarkTheme() and colors array to highlighter

CieNTi committed:
Added setDarkTheme() to datadisplay

CieNTi committed:
darkTheme set chain trigger. Light and dark coexistence

CieNTi committed:
Merge branch 'feature/add-dark-theme' into develop

CieNTi committed:
Added monospace option to settings and cpanel

CieNTi committed:
Added setMonoFont() to highlighter

CieNTi committed:
Added setMonoFont() to datadisplay

CieNTi committed:
monoFont set chain trigger. Default/Mono works both on Light/Dark

CieNTi committed:
Merge branch 'feature/add-monospace-font' into develop

CieNTi committed:
Added font size option to settings and cpanel

CieNTi committed:
Added setFontSize() to highlighter

CieNTi committed:
Added setFontSize() to datadisplay

CieNTi committed:
fontSize set chain trigger. Working on Default/Mono font and Light/Dark

CieNTi committed:
Merge branch 'feature/add-font-size' into develop

CieNTi committed:
Update changelog and credits

CieNTi committed:
Merge branch 'feature/add-contrib' into develop

CieNTi committed:
Merge branch 'develop'

CieNTi committed:
Removing uneeded debug

CieNTi committed:
Fix exec type (Debian did not recognize it as program, still executable)

Extended info: https://forum.qt.io/topic/80690/solved-debian-9-qt-program-compile/12

CieNTi committed:
Merge branch 'feature/fix_x-sharedlib_debian' into develop

CieNTi committed:
Merge branch 'develop'

CieNTi commented:
Added fix for 'x-sharedlib' issue. See: https://forum.qt.io/topic/80690/solved-debian-9-qt-program-compile/12

CieNTi committed:
Removed wrong line

CieNTi committed:
Merge branch 'develop'

CieNTi commented:
*facepalm* .. enough for today

cyc1ingsir commented on CuteCom.pro:
```diff
@@ -58,9 +58,12 @@ FORMS    += mainwindow.ui \
 DISTFILES += \
     qt.astylerc
 
+QMAKE_LFLAGS += -no-pie
```
This line prevents from compiling when trying to build it with Qt Creator. (Unknown option)

CieNTi replied:
Hi! Please can you add more information about this? Because I'm compiling it right now using Qt Creator on Linux. Please explain me your test process so I can repeat it.

Looking the link I provided, it's Qt's gcc related flag. It is your build chain using a different compiler?

Anyway, I set this flag to fix a problem where executables are not recognized (double click from nemo launch the 'Open with ..' dialog instead of executing it, because is recognized as shared library instead executable, see the link i posted before for expanded information). Do you know a different fix?

Thanks

CieNTi replied:
I found this: https://stackoverflow.com/questions/45329372/ubuntu-recognizes-executable-as-shared-library-and-wont-run-it-by-clicking and tried to analyse in depth with CuteCom:

1. Only Debian-like distros will suffer this because gcc defaults on "-pie" flag ON, so "-no-pie" fix it
2. It seems only when using gcc and only at linking phase, others compilers are not affected
3. If set on .pro file, CMake still generates a library-like executable
4. If set on CMakeLists.txt, Qt Creator is not aware of this so a library-like executable is generated

Starting with Debian 9 and similar distros, default CuteCom executable is not correct: Can run inside Qt Creator and from command-line, but never by double-clicking it. I tested it compiling a fresh repository clone both from command line cmake+make and Qt Creator (default gcc kit ON). Reason is discussed on the link I provided.

To fix the incompatibilities you mentioned, I removed it from repository, but **how to deal with this?** A Debian user will not get a correctly composed executable, and needs to pass a flag to cmake / qt creator. **How to do this automatically?** Any idea?

cyc1ingsir commented on main.cpp:
```diff
@@ -47,6 +47,22 @@ int main(int argc, char *argv[])
+
+    /* Apply dark stylesheet if selected */
+    if (session_config.darkTheme) {
+        //QFile file(":/res/styles/style.qss");
```
This comment could be removed.

CieNTi replied:
Oops .. true

cyc1ingsir reviewed (changes requested):
Hi, many thanks for your PR. The screenshots look nice. 

Unfortunately this doesn't work when compiled using cmake.
The style needs to be added to the CMakeLists.txt, i guess.
Also, I'm getting some warnings when running make:
```
In file included from datadisplay.cpp:26:0:
datadisplay.h: In constructor 'DataDisplay::DataDisplay(QWidget*)':
datadisplay.h:136:9: warning: 'DataDisplay::m_fontSize' will be initialized after [-Wreorder]
     int m_fontSize;
         ^
datadisplay.h:121:10: warning:   'char DataDisplay::m_linebreakChar' [-Wreorder]
     char m_linebreakChar;
          ^
datadisplay.cpp:36:1: warning:   when initialized here [-Wreorder]
 DataDisplay::DataDisplay(QWidget *parent)
 ^
```

Could you please change the code accordingly.

How much of the Python code within the style is really needed?
Could some of it been removed?

Finally.
The Settings panel is meant to provide quick access to the settings of the line parameters and settings which the user is likely to change more often (e.g. Name of the logfile).
Adding the checkbox for the theme and font (size) leads to a overly crowed UI quickly.
So we should consider adding a settings dialog.
As a bonus, it would be even nicer if settings for the different fonts being used could be provided.

cyc1ingsir commented:
Something I forgot:
Running `for x in `ls *.cpp`; do ./format.sh $x; done` to check the code style leads to a few differences.
Could you please have a look.  
`for x in `ls *.h`; do ./format.sh $x; done` will fix a minor code style error in one of the header files. Please have a look.

diggit commented:
Why do you want to hard-code theme? If user wants dark theme, he probably has modified global color palette... (like me)

CieNTi commented:
@diggit If I understand well, and you meant to change system pallete, maybe the user wants to keep everything as is and only set a darker cutecom (like me). If you meant to edit cutecom code and recompile it just to have a darker color, maybe is an extra work not everybody wants to do (like me, too xD). If there is an option, it is just plain easier to set it.

Anyway If you pay attention to my changes, I just took a stylesheet and apply it. Every hardcoded color you see was hardcoded at first, I just compound it on groups (see datadisplay.cpp changes, for an example) where is easier to maintain. Is not perfect but it is a first approach, tell me a better idea and i'll apply it.

But here is a question about this: How do you retrieve a color from the stylesheet? I'll always prefer that to explicitily set it on code, if someone tells me that and I'll set it everything not-hardcoded, as it should be.

Thanks

CieNTi commented:
@cyc1ingsir About the theme code (and python question): I used https://github.com/ColinDuquesnoy/QDarkStyleSheet and I think it's preferable to add it as submodule, but I don't know the CuteCom strategy about this method, so I ended adding it as part of the repository itself.

If you think a submodule is a better approach to keep every part in its place, tell me and I modify it accordingly. But about the explicit Python question, I don't really know because I do not use it, i just took a guy's work and add it here "as is".

CieNTi committed:
Update CMakeLists Qt5 library finder

See: https://gitlab.kitware.com/cmake/cmake/merge_requests/226/diffs

diggit commented:
@CieNTi 
Yes, I like apps that respect "system" palette. If this possibility remains, I am ok with that :)

CieNTi committed:
Remove warning about reorder

CieNTi committed:
No more warnings about unused variables

CieNTi committed:
Removed not needed comment

CieNTi committed:
Dark style is correctly added both in Qt Creator and CMake

CieNTi committed:
Removed previously added '-no-pie' incompatible linker flag

Debian users still need to add '-no-pie' option to linker flags to get a valid executable, instead of a 'x-sharedlib' object.

CieNTi committed:
Merge branch 'feature/pr-fixes' into develop

CieNTi committed:
Merge branch 'develop'

cyc1ingsir commented:
Even though, it's a nice idea really would love to have it as a dependency. However, I've been told adding dependencies is not easy (or impossible?) with CMake. 
A submodule might be indeed a valid approach.
I'm closing this PR for now since CuteCom's repo has been moved.
Thank you very much anyway.
